### PR TITLE
Add SDK Sampling interface

### PR DESCRIPTION
**Pull request by joshua-kim**
## Why this should be merged

Allow the SDK clients to configure their node sampling strategy for `AppRequestAny`.

## How this works

Moves the peer state in `Router` (which was questionable anyways) to an implementation of an interface. We now expose two different implementations, one to sample any peer and one to sample strictly validators. The VM is expected to keep a reference to a set of samplers, and re-use them for any clients it might create.

## How this was tested

Added some unit tests.

### CODE REVIEW

File: network/p2p/client.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ type Client struct {
 	handlerPrefix []byte
 	router        *Router
 	sender        common.AppSender
+	nodeSampler   NodeSampler
```
Comment (joshua-kim) [inline network/p2p/client.go:44]: So I could potentially see some confusion where we now expose an interface that allows us to choose our sampling behavior. This is distinct from `AppGossip`, which still uses the config passed when the underlying `sender` was created.

File: network/p2p/node_sampler.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+func (v Validators) Sample(n int) ([]ids.NodeID, error) {
+	currentHeight, err := v.Validators.GetCurrentHeight(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current height: %w", err)
+	}
+
+	validatorSet, err := v.Validators.GetValidatorSet(context.TODO(), currentHeight, v.SubnetID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current validator set: %w", err)
+	}
```
Comment (StephenButtolph) [inline network/p2p/node_sampler.go:48]: This should be cached - we don't need exact validator sets - so maybe update it every few minutes max

File: network/p2p/router.go [hunk 3]
```diff
@@ -56,26 +51,10 @@ func NewRouter(log logging.Logger, sender common.AppSender) *Router {
...
-	return nil
-}
-
-func (r *Router) Disconnected(_ context.Context, nodeID ids.NodeID) error {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-
-	r.peers.Remove(nodeID)
-	return nil
-}
```
Comment (hexfusion) [inline network/p2p/router.go:73]: +1

File: network/p2p/router_test.go [hunk 4]
```diff
@@ -192,9 +191,9 @@ func TestAppRequestResponse(t *testing.T) {
 			sender := common.NewMockSender(ctrl)
 			handler := mocks.NewMockHandler(ctrl)
 			router := NewRouter(logging.NoLog{}, sender)
-			require.NoError(router.Connected(context.Background(), nodeID, nil))
-
-			client, err := router.RegisterAppProtocol(handlerID, handler)
+			sampler := &Peers{}
```
Comment (hexfusion) [inline network/p2p/router_test.go:194]: this feels strange its clearly `peers` I get that it implements `Sample` is this `peerSampler`?

File: network/p2p/router.go [hunk 5]
```diff
@@ -56,26 +51,10 @@ func NewRouter(log logging.Logger, sender common.AppSender) *Router {
...
-
 // RegisterAppProtocol reserves an identifier for an application protocol and
 // returns a Client that can be used to send messages for the corresponding
 // protocol.
-func (r *Router) RegisterAppProtocol(handlerID uint64, handler Handler) (*Client, error) {
+func (r *Router) RegisterAppProtocol(handlerID uint64, handler Handler, nodeSampler NodeSampler) (*Client, error) {
```
Comment (hexfusion) [inline network/p2p/router.go:57]: We want to be careful about breaking SDK changes. Probably not worth changing now but if you foresee more configurations for the protocols we register might be nice to make this a variadic func or take a config.
```suggestion
(r *Router) RegisterAppProtocol(handlerID uint64, handler Handler, opts ...ProtocolOption) (*Client, error) {
```

File: network/p2p/node_sampler_test.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+			sampler := &Peers{}
+
+			for connected := range tt.connected {
+				require.NoError(sampler.Connected(context.Background(), connected, nil))
+			}
+
+			for disconnected := range tt.disconnected {
+				require.NoError(sampler.Disconnected(context.Background(), disconnected))
```
Comment (hexfusion) [inline network/p2p/node_sampler_test.go:90]: why not use context.WithCancel for the test?

File: network/p2p/node_sampler.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	_ NodeSampler = (*Validators)(nil)
+
+	_ validators.Connector = (*Peers)(nil)
+	_ NodeSampler          = (*Peers)(nil)
+)
+
+type NodeSampler interface {
+	// Sample returns at most [n] nodes. This may return fewer nodes if fewer
+	// than [n] are available.
+	Sample(n int) ([]ids.NodeID, error)
```
[the remark this answers is not in the document]
Comment (hexfusion) [reply network/p2p/node_sampler.go:36]: Can you give an example of a sample that would return an error?

File: network/p2p/node_sampler.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+var (
+	_ NodeSampler = (*Validators)(nil)
+
+	_ validators.Connector = (*Peers)(nil)
+	_ NodeSampler          = (*Peers)(nil)
+)
+
+type NodeSampler interface {
```
Comment (hexfusion) [inline network/p2p/node_sampler.go:33]: Is this `NodeSampler` or `PeerSampler`

File: network/p2p/node_sampler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/node_sampler.go:36]: > Shouldn't Sample take a context? You are already needing to stub out TODO getRecentValidatorIDs.

Yeah this makes sense, I can add this to the signature

> Can you give an example of a sample that would return an error?

Hmm I guess an example would be some grpc failure? Some implementation might hit the p-chain for example

File: network/p2p/node_sampler_test.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+func TestPeersSample(t *testing.T) {
+	nodeID1 := ids.GenerateTestNodeID()
+	nodeID2 := ids.GenerateTestNodeID()
+	nodeID3 := ids.GenerateTestNodeID()
+
+	tests := []struct {
+		name         string
+		connected    set.Set[ids.NodeID]
+		disconnected set.Set[ids.NodeID]
+		n            int
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/node_sampler_test.go:29]: Hmm this example test case you provided should be valid, `n` is supposed to be a limit parameter to limit the output being sampled. I can change name here.

File: network/p2p/router_test.go (hunk 4 above)
Comment (joshua-kim) [reply network/p2p/router_test.go:194]: I can just rename this to `peers` or `peerSampler`.

File: network/p2p/node_sampler_test.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	safemath "github.com/ava-labs/avalanchego/utils/math"
```
Comment (StephenButtolph) [inline network/p2p/node_sampler_test.go:16]:
```suggestion
	"github.com/ava-labs/avalanchego/utils/math"
```

File: network/p2p/node_sampler_test.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+package p2p
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
```
Comment (StephenButtolph) [inline network/p2p/node_sampler_test.go:12]:
```suggestion
	"github.com/stretchr/testify/require"

	"go.uber.org/mock/gomock"
```

File: network/p2p/node_sampler.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current height: %w", err)
+	}
+
+	validatorSet, err := v.validators.GetValidatorSet(context.TODO(), height, v.subnetID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current validator set: %w", err)
+	}
+
+	v.recentValidatorIDs = maps.Keys(validatorSet)
```
Comment (StephenButtolph) [inline network/p2p/node_sampler.go:71]: This is expected to be called concurrently right? This function needs a lock

File: network/p2p/node_sampler.go (hunk 8 above)
Comment (StephenButtolph) [reply network/p2p/node_sampler.go:33]: Seems more like a node sampler to me. It samples nodes (which may or may not be connected)

Comment (joshua-kim) [reply network/p2p/node_sampler.go:33]: Yeah my thinking was that the terminology we use are:
- `node`: a member in the network
- `peer`: a node we are connected to
- `validator`: a staking node

File: network/p2p/router.go (hunk 5 above)
Comment (joshua-kim) [reply network/p2p/router.go:57]: The options idea is interesting. GRPC seems to have a private `apply` interface to apply the configs. I think something like that seems interesting, I'll look into it.

Comment (joshua-kim) [reply network/p2p/router.go:57]: Hmm I don't want to add a variadic options field yet, I think it makes this more complex. I think it's something we definitely want to add later though. We probably don't have any users for this package yet so I'm fine making a breaking change here.

My main concern is that if we make this an options field we need to figure out a way to inject a default node sampler that the option would override. The only way to do this is if we add a `Peers` or `Validators` to the `Router`, which I feel like looks strange since I just want the `Router` to have the responsibility of handling message routing logic.

File: network/p2p/node_sampler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/node_sampler.go:36]: Actually, the only reason we have this `Sample` interface is to use it for the `Client` node selection implementation. We could actually swallow this error and just return an empty slice of node ids, and have the `Client` report an error if it sees no nodes to send to, so the error is still propagated and this interface signature is much cleaner since we can get rid of the error here.

File: network/p2p/node_sampler_test.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+			require.Subset(sampleable, sampled)
+		})
+	}
+}
+
+func TestValidatorsSample(t *testing.T) {
+	tests := []struct {
+		name       string
+		validators []ids.NodeID
+		n          int
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/node_sampler_test.go:110]: Added an error case

File: network/p2p/node_sampler_test.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+			validatorSet := make(map[ids.NodeID]*validators.GetValidatorOutput, 0)
+			for _, validator := range tt.validators {
+				validatorSet[validator] = nil
+			}
+			mockValidators.EXPECT().GetValidatorSet(gomock.Any(), height, subnetID).Return(validatorSet, nil)
+
+			v := NewValidators(subnetID, mockValidators)
+
+			sampled, err := v.Sample(tt.n)
+			require.NoError(err)
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/node_sampler_test.go:156]: See previous comment reply, I think we can remove the error here.

File: network/p2p/router.go (hunk 5 above)
Comment (hexfusion) [reply network/p2p/router.go:57]: just something to think about for now

File: network/p2p/node_sampler_test.go (hunk 6 above)
Comment (joshua-kim) [reply network/p2p/node_sampler_test.go:91]: Hum can you elaborate? I didn't understand why this would be helpful here.

File: network/p2p/node_sampler.go [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+	"context"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+// NodeSampler samples nodes in network
+type NodeSampler interface {
+	// Sample returns at most [limit] nodes. This may return fewer nodes if
+	// fewer than [limit] are available.
+	Sample(ctx context.Context, limit int) []ids.NodeID
```
Comment (joshua-kim) [inline network/p2p/node_sampler.go:16]: note: I also considered making another type called:
```
type Node struct {
    ID ids.NodeID
}
```

so we could add onto it without causing breaking changes. I felt like it was reasonable that in the future we might add some peer-level metadata to this, but it seemed premature to add for now.

File: network/p2p/validators.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+const maxValidatorSetStaleness = time.Minute
```
Comment (danlaine) [inline network/p2p/validators.go:16]: Should we make this configurable?

File: network/p2p/validators.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+	height, err := v.validators.GetCurrentHeight(ctx)
+	if err != nil {
+		return
+	}
+	validatorSet, err := v.validators.GetValidatorSet(ctx, height, v.subnetID)
+	if err != nil {
+		return
+	}
+
+	v.validatorIDs = set.NewSampleableSet[ids.NodeID](len(validatorSet))
```
Comment (danlaine) [inline network/p2p/validators.go:53]: Could also do `v.validatorIDs.Clear` to prevent re-allocation

File: network/p2p/validators.go [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	height, err := v.validators.GetCurrentHeight(ctx)
+	if err != nil {
+		return
+	}
+	validatorSet, err := v.validators.GetValidatorSet(ctx, height, v.subnetID)
+	if err != nil {
+		return
+	}
```
Comment (danlaine) [inline network/p2p/validators.go:51]: Hmm is there something we could do other than swallow these errors? Adding `error` return values seems kind of annoying but also I dislike swallowing errors. Thoughts?

File: network/p2p/client.go [hunk 19]
```diff
@@ -41,6 +41,7 @@ type Client struct {
 	handlerPrefix []byte
```
Comment (danlaine) [inline network/p2p/client.go:41]: Not part of this PR but should we have a `NewClient` function? If this in intended to be used only inside this package, `Client` can be unexported. If it's intended to be used outside this package, then a user can't use `Client` because they can't set its inner fields.

File: network/p2p/client.go (hunk 1 above)
Comment (danlaine) [reply network/p2p/client.go:44]: Comments never hurt :)

Comment (danlaine) [reply network/p2p/client.go:44]: Or we could have `AppGossip` use the `NodeSampler` -- that would seem reasonable to me

File: network/p2p/client.go (hunk 19 above)
Comment (joshua-kim) [reply network/p2p/client.go:41]: This is created via `RegisterAppProtocol`. It's not possible to use a `Client` without a corresponding `Router` because the `Router` holds the map of pending requests -> the callbacks that need to be fired on response. We could theoretically refactor this into a `NewClient(Router, handlerID, ....)` if we felt that looked cleaner.

File: network/p2p/client.go (hunk 1 above)
Comment (joshua-kim) [reply network/p2p/client.go:44]: issue: https://github.com/ava-labs/avalanchego/issues/1890

Comment (danlaine) [reply network/p2p/client.go:44]: In that case we should probably put a comment explaining when `nodeSampler` is used and when the sender's config is used

File: network/p2p/client.go (hunk 19 above)
Comment (danlaine) [reply network/p2p/client.go:41]: It just feels weird to export this if it's actually unusable outside this package

File: network/p2p/node_sampler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (danlaine) [reply network/p2p/node_sampler.go]: > We could actually swallow this error and just return an empty slice of node ids, and have the Client report an error if it sees no nodes to send to,

If we error and return on calls to `v.validators.GetCurrentHeight(ctx)` or `v.validators.GetValidatorSet(ctx, height, v.subnetID)` we won't update `v.validatorIDs` so future sampling won't return an empty slice of node IDs, it'll just return stale node IDs. FWIW the sampler implementations in the `sampler` package have errors in their return signature.

File: network/p2p/client.go (hunk 19 above)
Comment (joshua-kim) [reply network/p2p/client.go:41]: `Client` is usable outside of this package, `RegisterAppProtocol` initializes it [here](https://github.com/ava-labs/avalanchego/blob/499a7f2334166919addd7ce7b00d59c92dfba6d4/network/p2p/router.go#L72).

Maybe we could rename this to `NewClient` to be more verbose or remove the receiver on `Router`. This definitely might seem confusing coming from the existing paradigm of how we've developed VMs without the SDK component.

File: network/p2p/validators.go (hunk 16 above)
Comment (joshua-kim) [reply network/p2p/validators.go:16]: Originally didn't make this configurable because I wanted to remove this whole refresh logic and not cause a breaking change. My concern was probably premature since I don't think anyone is using SDK at the moment, so I will make this configurable.

File: network/p2p/client.go (hunk 19 above)
Comment (danlaine) [reply network/p2p/client.go:41]: Ah I see. Yeah I think it might be a bit confusing to someone looking at this package how to create a new client. I think maybe even just a comment on `Client` might suffice.

Comment (joshua-kim) [reply network/p2p/client.go:41]: Yeah I could see an argument to make the constructor code `NewClient(sender common.AppSender, router *Router, handlerID byte) *Client`. This would have the added benefit of not having to keep around the `sender` as a field in `Router` which we currently do.

File: network/p2p/node_sampler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/node_sampler.go]: I added a unit test for this specific scenario to make sure we return an empty slice instead of stale ids so the error is surfaced correctly in the `Client`.

**Review by danlaine — APPROVED** [review]

File: network/p2p/validators_test.go [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+package p2p
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
```
Comment (StephenButtolph) [inline network/p2p/validators_test.go:13]:
```suggestion
	"github.com/stretchr/testify/require"

	"go.uber.org/mock/gomock"
```

**Review by StephenButtolph — APPROVED** [review]

**Review by darioush — APPROVED** [review]
